ext/standard/php_uuid.h

```diff
@@ -0,0 +1,163 @@
+  | This source file is subject to version 3.01 of the PHP license,      |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.php.net/license/3_01.txt                                  |
+  | If you did not receive a copy of the PHP license and are unable to   |
+  | obtain it through the world-wide-web, please send a note to          |
+  | license@php.net so we can mail you a copy immediately.               |
+  +----------------------------------------------------------------------+
+  | Author: Richard Fussenegger <php@fleshgrinder.com>                   |
+  +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#ifndef PHP_UUID_H
+#define PHP_UUID_H
+
+#include "zend_types.h"
+#include "php.h"
+
+PHPAPI extern zend_class_entry *php_ce_UUID;
+PHPAPI extern zend_class_entry *php_ce_UUIDParseException;
+
+#define PHP_UUID_LEN 16
+typedef struct php_uuid {
+	uint8_t bytes[PHP_UUID_LEN];
+} php_uuid;
+
+#define PHP_UUID_HEX_LEN 33
+typedef struct php_uuid_hex {
+	char str[PHP_UUID_HEX_LEN];
+} php_uuid_hex;
+
+#define PHP_UUID_STRING_LEN 37
+typedef struct php_uuid_string {
+	char str[PHP_UUID_STRING_LEN];
+} php_uuid_string;
+
+typedef enum php_uuid_variant {
+	PHP_UUID_VARIANT_NCS             = 0,
+	PHP_UUID_VARIANT_RFC4122         = 1,
+	PHP_UUID_VARIANT_MICROSOFT       = 2,
+	PHP_UUID_VARIANT_FUTURE_RESERVED = 3,
+} php_uuid_variant;
+
+static const uint8_t PHP_UUID_VERSION_1_TIME_BASED      = 1;
+static const uint8_t PHP_UUID_VERSION_2_DCE_SECURITY    = 2;
+static const uint8_t PHP_UUID_VERSION_3_NAME_BASED_MD5  = 3;
+static const uint8_t PHP_UUID_VERSION_4_RANDOM          = 4;
+static const uint8_t PHP_UUID_VERSION_5_NAME_BASED_SHA1 = 5;
+
+BEGIN_EXTERN_C()
+
+PHPAPI void php_uuid_create_v3(php_uuid *uuid, const php_uuid *namespace, const char *name, const size_t name_len);
+
+PHPAPI int php_uuid_create_v4(php_uuid *uuid, const zend_bool throw);
+
+static zend_always_inline int php_uuid_create_v4_silent(php_uuid *uuid)
+{
+	return php_uuid_create_v4(uuid, 0);
+}
+
+static zend_always_inline int php_uuid_create_v4_throw(php_uuid *uuid)
+{
+	return php_uuid_create_v4(uuid, 1);
+}
+
+PHPAPI void php_uuid_create_v5(php_uuid *uuid, const php_uuid *namespace, const char *name, const size_t name_len);
+
+static const zend_always_inline php_uuid_variant php_uuid_get_variant(const php_uuid *uuid)
+{
+	if ((uuid->bytes[8] & 0xC0) == 0x80) return PHP_UUID_VARIANT_RFC4122;
+	if ((uuid->bytes[8] & 0xE0) == 0xC0) return PHP_UUID_VARIANT_MICROSOFT;
+	if ((uuid->bytes[8] & 0x80) == 0x00) return PHP_UUID_VARIANT_NCS;
+	return PHP_UUID_VARIANT_FUTURE_RESERVED;
+}
+
+static const zend_always_inline uint8_t php_uuid_get_version(const php_uuid *uuid)
+{
+	return uuid->bytes[6] >> 4;
+}
+
+static const zend_always_inline php_uuid php_uuid_namespace_dns()
```

Comment:
Shouldn't `const` here come before `php_uuid`? `const` is a property of the type `php_uuid`, not of the function itself, right?

Reply:
Not according to the gcc documentation, it would also make no sense. How could the function ensure that the returned value stays constant?

> Many functions do not examine any values except their arguments, and have no effects except the return value. Basically this is just slightly more strict class than the `pure` attribute below, since function is not allowed to read global memory.
>
> Note that a function that has pointer arguments and examines the data pointed to must not be declared `const`. Likewise, a function that calls a non-`const` function usually must not be `const`. It does not make sense for a `const` function to return `void`.
>
> — [gcc Common Function Attributes](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html)

Reply:
Disregard my comment, @nikic said that this documentation is about `__attribute__((const))` and not about `const`.

Reply:
also I'm not sure why `const` should be used when you're returning a struct

Reply:
Well, I thought that this is the `const` attribute that tells the compiler that this function will always return the same data structure; which it does. 😉 However, turns out that this one is different.